### ISSUE TITLE
Update `styled-components` dependencies

### ISSUE THREAD

**Issue by dgreif** [primer/react#1298]
I recently updated most of the base-level dependencies (https://github.com/primer/components/pull/1292).  One that I elected to skip was `styled-components`, which has a [major release with big perf improvements](https://medium.com/styled-components/announcing-styled-components-v5-beast-mode-389747abd987).  I tried updating, but most of the snapshots needed to be updated and there were other issues with the tests.  I don't think the changes are actually significantly breaking, but this bump should be handled separate from other dependency bumps so that it gets the proper attention it needs.

**Comment by moatorres** [comment]
Hey @dgreif, 

First of all, thank you for making Primer so awesome! Primer ❤️ 

I've been experimenting with `@primer/components` for the past few days and I've found a weird bug (possibly related to `ThemeProvider` component) that creates unexpected behaviors (screenshot attached). 

In this particular case it's messing with how the styles are applied when switching between color modes. The `selected` `SideNav.Link` and the `color` of all items are "freezing", as well as all `:hover` styles not being applied they should.

### TL;DR

I strongly advise against updating `styled-components` to anything above `5.1.1`. I'd also suggest to change primer's `peerDependencies` as below.

```js
"peerDependencies": {
    "react": "^17.0.0",
    "react-dom": "^17.0.0",
    "styled-components": "4.x || 5.1.x" // or even "^5.1.x"
 }
```

<details><summary>Steps to reproduce</summary>

1. Update `styled-components` in `package.json`
```js
"devDependencies": {
   ...
  "styled-components": "5.2.0" // or any version above
}
```
2. Run `npm run playroom:start`
3. Add the snippet below (from [primer.style](https://primer.style/components/SideNav) examples)
```jsx
<ThemeProvider colorMode="night">
  <SideNav bordered maxWidth={360} aria-label="Main" m={4}>
    <SideNav.Link href="#account">
      <Text>Account</Text>
    </SideNav.Link>
    <SideNav.Link href="#home" selected>
      <Text>Profile</Text>
    </SideNav.Link>
    <SideNav.Link href="#emails">
      <Text>Emails</Text>
    </SideNav.Link>
    <SideNav.Link href="#notifications">
      <Text>Notifications</Text>
    </SideNav.Link>
  </SideNav>
</ThemeProvider>
```
4. Change `colorMode="night"` to `colorMode="day"`
5. Hit `Ctrl (Cmd) + Z` and `Ctrl (Cmd) + Shift + Z` to toggle the `colorMode` (you'll get the same behavior using `setColorMode` from `useTheme` or any `useState` method to update the `colorMode` from `day` to `night`, etc, regardless of which `dayScheme` or `nightScheme` is set)

I've also got the same results adding `@primer/components` to a project that has `styled-components@>=5.2.x` installed.

![primer-sidenav](https://user-images.githubusercontent.com/44585769/124700150-99c86180-dec2-11eb-994b-4e2252f46bc7.jpg)

</details>

I didn't get the time to run through `styled-components` changelogs yet, but I've spent a good couple of hours to find out what was going on. I can confirm that everything works normally up to version `5.1.1` and that above that will bug users' heads.

**Comment by moatorres** [comment]
Hey @dgreif, 

First of all, thank you for making Primer so awesome! ❤️ 

I've found a weird bug (possibly related to the `ThemeProvider` component) that creates unexpected behaviors (screenshot attached). It's messing with how styles are applied when switching between color modes. The `selected` `SideNav.Link` and the `color` of  any`SideNav.Item` are "freezing", as well as `:hover` styles.

### TL;DR

I strongly advise against updating `styled-components` to anything above `5.1.1`. I'd also suggest to change primer's `peerDependencies` as below.

```js
"peerDependencies": {
    "react": "^17.0.0",
    "react-dom": "^17.0.0",
    "styled-components": "4.x || 5.1.x" // or even "^5.1.x"
 }
```

<details><summary>Steps to reproduce</summary>

1. Update `styled-components` in `package.json`
```js
"devDependencies": {
   ...
  "styled-components": "5.2.0" // or any version above
}
```
2. Run `npm run playroom:start`
3. Add the snippet below (from [primer.style](https://primer.style/components/SideNav) examples)
```jsx
<ThemeProvider colorMode="night">
  <SideNav bordered maxWidth={360} aria-label="Main" m={4}>
    <SideNav.Link href="#account">
      <Text>Account</Text>
    </SideNav.Link>
    <SideNav.Link href="#home" selected>
      <Text>Profile</Text>
    </SideNav.Link>
    <SideNav.Link href="#emails">
      <Text>Emails</Text>
    </SideNav.Link>
    <SideNav.Link href="#notifications">
      <Text>Notifications</Text>
    </SideNav.Link>
  </SideNav>
</ThemeProvider>
```
4. Change `colorMode="night"` to `colorMode="day"`
5. Hit `Ctrl (Cmd) + Z` and `Ctrl (Cmd) + Shift + Z` to toggle the `colorMode` (you'll get the same behavior using `setColorMode` from `useTheme` or any `useState` method to update the `colorMode` from `day` to `night`, etc, regardless of which `dayScheme` or `nightScheme` is set)

I've got the same results adding `@primer/components` to a project that has `styled-components@>=5.2.x` installed.

![primer-sidenav](https://user-images.githubusercontent.com/44585769/124700150-99c86180-dec2-11eb-994b-4e2252f46bc7.jpg)

</details>

I didn't get the time to run through `styled-components` changelogs yet, but I've spent a good couple of hours to find out what happened. I can confirm that everything works normally up to version `5.1.1` and that anything  `>=5.2.x` will bug users' heads.

**Comment by moatorres** [comment]
Hey @dgreif, 

First of all, thank you for making Primer so awesome! ❤️ 

I've found a weird bug (possibly related to the `ThemeProvider` component) that creates unexpected behaviors (screenshot attached). It's messing with how styles are applied when switching between color modes. The `selected` `SideNav.Link` and the `color` of  any`SideNav.Item` are "freezing", as well as `:hover` styles.

### TL;DR

I strongly advise against updating `styled-components` to anything above `5.1.1`. I'd also suggest to change primer's `peerDependencies` as below.

```js
"peerDependencies": {
    "react": "^17.0.0",
    "react-dom": "^17.0.0",
    "styled-components": "4.x || 5.1.x" // or even "^5.1.x"
 }
```

<details><summary>Steps to reproduce</summary>

1. Update `styled-components` in `package.json`
```js
"devDependencies": {
   ...
  "styled-components": "5.2.0" // or any version above
}
```
2. Run `npm run playroom:start`
3. Add the snippet below (from [primer.style](https://primer.style/components/SideNav) examples)
```jsx
<ThemeProvider colorMode="night">
  <SideNav bordered maxWidth={360} aria-label="Main" m={4}>
    <SideNav.Link href="#account">
      <Text>Account</Text>
    </SideNav.Link>
    <SideNav.Link href="#home" selected>
      <Text>Profile</Text>
    </SideNav.Link>
    <SideNav.Link href="#emails">
      <Text>Emails</Text>
    </SideNav.Link>
    <SideNav.Link href="#notifications">
      <Text>Notifications</Text>
    </SideNav.Link>
  </SideNav>
</ThemeProvider>
```
4. Change `colorMode="night"` to `colorMode="day"`
5. Hit `Ctrl (Cmd) + Z` and `Ctrl (Cmd) + Shift + Z` to toggle the `colorMode` (you'll get the same behavior using `setColorMode` from `useTheme` or any `useState` method to update the `colorMode` from `day` to `night`, etc, regardless of which `dayScheme` or `nightScheme` is set)

I've got the same issue using `@primer/components` on a project that had `styled-components@5.3.0` installed.

![primer-sidenav](https://user-images.githubusercontent.com/44585769/124700150-99c86180-dec2-11eb-994b-4e2252f46bc7.jpg)

</details>

I didn't get the time to run through `styled-components` changelogs yet, but I've spent a good couple of hours to find out what happened. I can confirm that everything works normally up to version `5.1.1` and that anything  `>=5.2.x` will bug users' heads.

**Comment by moatorres** [comment]
My best guess is that the patch below is the origin of the issue.

From [sc releases](https://styled-components.com/releases):
> ### v5.2.0-test.10 + v5.2.0
> Rework how components self-reference in extension scenarios (see #3236); should fix a bunch of subtle bugs around patterns like & + &

You might want to check issues [3236](https://github.com/styled-components/styled-components/pull/3236), [3265](https://github.com/styled-components/styled-components/issues/3265) and [3401](https://github.com/styled-components/styled-components/issues/3401) from `styled-components` repo.

**Comment by moatorres** [comment]
And this reproduction here: https://codesandbox.io/s/styled-components-parent-selector-bug-v54si?file=/package.json ✨

**Comment by colebemis** [comment]
Looks like there is an open issue with `jest-styled-components` (https://github.com/styled-components/jest-styled-components/issues/276) that breaks our snapshot tests when we upgrade to v5. 

I'd argue that this is a low priority for us to fix right now because, since styled-components is a peer dependency, consumers of Primer React can use v5 in their application without any changes on the Primer React side (Memex already uses v5).

**Comment by colebemis** [comment]
Closing this for now until we determine if we want to use styled-components long term: https://github.com/github/primer/issues/203
